tasks/main.yml

```diff
@@ -1,32 +1,154 @@
+    - '{{ apt__packages }}'
+  when: apt__enabled|bool
+
+- name: Ensure that APT default configuration is absent
+  file:
+    path: '/etc/apt/apt.conf'
+    state: 'absent'
+  when: (apt__enabled|bool and apt__remove_default_configuration|bool)
+
+- name: Generate APT configuration files
+  template:
+    src: '{{ item }}.j2'
+    dest: '/{{ item }}'
+    owner: 'root'
+    group: 'root'
+    mode: '0644'
+  with_flattened:
+    - 'etc/apt/apt.conf.d/25no-recommends.conf'
+  when: apt__enabled|bool
+
+- name: Check current APT diversions
+  environment:
+    LC_ALL: 'C'
+  shell: dpkg-divert --list '/etc/apt/*.dpkg-divert' | awk '{print $NF}'
+  register: apt__register_diversions
+  when: apt__enabled|bool
+  always_run: True
+  changed_when: False
+
+- name: Divert original /etc/apt/sources.list
+  command: dpkg-divert --quiet --local --divert /etc/apt/sources.list.dpkg-divert --rename /etc/apt/sources.list
+  when: (apt__enabled|bool and apt__sources_deploy_state == 'present' and
+         '/etc/apt/sources.list.dpkg-divert' not in apt__register_diversions.stdout_lines)
+
+- name: Configure APT distribution sources
+  template:
+    src: 'etc/apt/sources.list.j2'
+    dest: '/etc/apt/sources.list'
+    owner: 'root'
+    group: 'root'
     mode: '0644'
-  when: (apt__enabled|d() | bool)
+  register: apt__register_sources_template
+  when: (apt__enabled|bool and apt__sources_deploy_state == 'present')
+
+- name: Remove diversion of original /etc/apt/sources.list
+  shell: rm -f /etc/apt/sources.list ; dpkg-divert --quiet --local --rename --remove /etc/apt/sources.list
+  args:
+    warn: False
+  register: apt__register_sources_diversion
+  when: (apt__enabled|bool and apt__sources_deploy_state == 'absent' and
+         '/etc/apt/sources.list.dpkg-divert' in apt__register_diversions.stdout_lines)
+
+- name: Configure custom APT keys
+  apt_key:
+    data:      '{{ item.data      | d(omit) }}'
+    file:      '{{ item.file      | d(omit) }}'
+    id:        '{{ item.id        | d(omit) }}'
+    keyring:   '{{ item.keyring   | d(omit) }}'
+    keyserver: '{{ item.keyserver | d(omit) }}'
+    url:       '{{ item.url       | d(omit) }}'
+    state:     '{{ item.state     | d("present"
+                                      if (item.distribution|d(apt__distribution) == apt__distribution and
+                                          item.distribution_release|d(apt__distribution_release) == apt__distribution_release)
+                                      else omit) }}'
+  with_flattened:
+    - '{{ apt__keys }}'
+    - '{{ apt__group_keys }}'
+    - '{{ apt__host_keys }}'
+    - '{{ apt__dependent_keys }}'
+  when: (apt__enabled|bool and (item.url|d() or item.data|d() or item.id|d() or item.file|d()) and
+         (item.state|d("present") == 'present' and
+          (item.distribution|d(apt__distribution) == apt__distribution and
+           item.distribution_release|d(apt__distribution_release) == apt__distribution_release)))
```

Comment:
The issue with `apt_key` module is that when the key is not specified by at least `id` parameter, role fails with an error about unknown key when the result is to set the key as absent. Any suggestions how to fix that in a different way than using the `when` condition?